telegram/ext/_fsm/machine.py

```diff
@@ -0,0 +1,200 @@
+"""This Module contains the FiniteStateMachine class and the built-in subclass SingleStateMachine.
+"""
+
+import abc
+import asyncio
+import contextlib
+import datetime as dtm
+import logging
+import time
+import weakref
+from collections import defaultdict, deque
+from collections.abc import AsyncIterator, Hashable, Mapping, MutableSequence, Sequence
+from types import MappingProxyType
+from typing import TYPE_CHECKING, Any, Generic, Literal, Optional, TypeVar, Union, overload
+
+from telegram.ext._fsm.states import State
+from telegram.ext._utils.types import JobCallback
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from telegram.ext import JobQueue
+
+_KT = TypeVar("_KT", bound=Hashable)
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+
+class StateInfo(Generic[_KT]):
+    def __init__(self: "StateInfo[_KT]", key: _KT, state: State, version: int) -> None:
+        self.key: _KT = key
+        self.state: State = state
+        self.version: int = version
+
+
+class FiniteStateMachine(abc.ABC, Generic[_KT]):
+    def __init__(self) -> None:
+        self._locks: MutableMapping[_KT, asyncio.Lock] = weakref.WeakValueDictionary()
+
+        # There is likely litte benefit for a user to customize how exactly the states are stored
+        # and accessed. So we make this private and only provide a read-only view.
+        self.__states: dict[_KT, tuple[State, int]] = defaultdict(
+            lambda: (State.IDLE, time.perf_counter_ns())
+        )
+        self._states = MappingProxyType(self.__states)
+
+        self.__job_queue: Optional[weakref.ReferenceType[JobQueue]] = None
+        self.__history: Mapping[_KT, MutableSequence[State]] = defaultdict(
```

Comment:
Why are these with two underscores.

Oh and we should make history optional/limit it so bots on a bigger scale doesn't lead to huge memory issues

Reply:
https://peps.python.org/pep-0008/#designing-for-inheritance

This makes name mangling being used.
TBF I don't strictly rely on pep8 in these questions, i.e. I don't choose Single underscore vs Dunder based on exclusively based "should a subclass be allowed to use this?". We have the stability which basically says that also single underscore should not be used by users unless it's documented.
My rule of thumb is using dunder when I want to highlight that something is really low-level.
The fact that dunder names get name-mangled is a nice bonus.